File: src/test/java/com/jloom/state/ProjectStateStoreTest.java
package com.jloom.state;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;

import java.nio.file.Path;
import java.time.Instant;

import static org.assertj.core.api.Assertions.assertThat;

class ProjectStateStoreTest {

    @TempDir
    Path tempDir;

    private final ProjectStateStore store = new ProjectStateStore();

    @Test
    void loadReturnsEmptyWhenNoStateFileExists() {
        assertThat(store.load(tempDir)).isEqualTo(ProjectState.empty());
    }

    @Test
    void saveThenLoadRoundTrips() {
        ProjectState seeded = ProjectState.empty()
                .withBasePackage("com.acme.demo")
                .withProjectName("acme")
                .withApplied(new AppliedModule("postgres", "1.0.0", Instant.parse("2026-01-01T00:00:00Z"), java.util.Map.of("db_name", "demo")));

        store.save(tempDir, seeded);
        ProjectState loaded = store.load(tempDir);

        assertThat(loaded.basePackage()).isEqualTo("com.acme.demo");
        assertThat(loaded.projectName()).isEqualTo("acme");
        assertThat(loaded.appliedIds()).containsExactly("postgres");
        assertThat(loaded.modules().get(0).version()).isEqualTo("1.0.0");
        assertThat(loaded.modules().get(0).answers()).containsEntry("db_name", "demo");
    }

    @Test
    void saveCreatesTheStateDirectoryIfMissing() {
        Path fresh = tempDir.resolve("brand-new-project");
        store.save(fresh, ProjectState.empty().withProjectName("brand-new-project"));

        assertThat(store.load(fresh).projectName()).isEqualTo("brand-new-project");
    }
}

File: bun/test/projectStateStore.test.ts
import { test } from "node:test";
import assert from "node:assert/strict";
import { mkdtempSync, rmSync } from "node:fs";
import os from "node:os";
import path from "node:path";
import {
  emptyState,
  loadState,
  saveState,
  withApplied,
  withBasePackage,
  withProjectName,
} from "../src/state/projectStateStore.ts";

function tempDir(): string {
  return mkdtempSync(path.join(os.tmpdir(), "jloom-state-"));
}

test("load returns empty state when no state file exists", () => {
  const dir = tempDir();
  try {
    assert.deepStrictEqual(loadState(dir), emptyState());
  } finally {
    rmSync(dir, { recursive: true, force: true });
  }
});

test("save then load round trips", () => {
  const dir = tempDir();
  try {
    let seeded = emptyState();
    seeded = withBasePackage(seeded, "com.acme.demo");
    seeded = withProjectName(seeded, "acme");
    seeded = withApplied(seeded, {
      id: "postgres",
      version: "1.0.0",
      appliedAt: "2026-01-01T00:00:00.000Z",
      answers: { db_name: "demo" },
    });

    saveState(dir, seeded);
    const loaded = loadState(dir);

    assert.strictEqual(loaded.basePackage, "com.acme.demo");
    assert.strictEqual(loaded.projectName, "acme");
    assert.deepStrictEqual(loaded.modules.map((m) => m.id), ["postgres"]);
    assert.strictEqual(loaded.modules[0]!.version, "1.0.0");
    assert.strictEqual(loaded.modules[0]!.answers.db_name, "demo");
  } finally {
    rmSync(dir, { recursive: true, force: true });
  }
});

test("save creates the state directory if missing", () => {
  const dir = tempDir();
  try {
    const fresh = path.join(dir, "brand-new-project");
    saveState(fresh, withProjectName(emptyState(), "brand-new-project"));
    assert.strictEqual(loadState(fresh).projectName, "brand-new-project");
  } finally {
    rmSync(dir, { recursive: true, force: true });
  }
});
